### PR TITLE
Remove use of 2to3

### PR DESCRIPTION
**Pull request by rahulporuri**
Use `six` instead

~WARNING : I used `modernize` to make automated changes to the codebase. I then added the recommended changes without much review. As such, this PR might include changes which are not absolutely necessary to the functionality of the package. If necessary, I can review the changes and remove the unnecessary/overzealous changes.~

### CODE REVIEW

**Comment by rahulporuri** [comment]
The failures are now Python 2 only. See job https://travis-ci.org/enthought/apptools/jobs/531612677

**Review by jwiggins — COMMENTED** [review]
I think the blanket use of `from __future__ import absolute_import` is of dubious usefulness.

Additionally, it might be worth the time to examine each of the files where `six.moves` is being used. Specifically, `map`, `zip`, and `range` might be OK as is.

File: apptools/template/impl/any_context_data_name_item.py [hunk 1]
```diff
@@ -122,13 +124,13 @@ def inputs_changed ( self ):
             filter = self.filter
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/template/impl/any_context_data_name_item.py]: looks like it. corrected.

File: apptools/persistence/state_pickler.py [hunk 2]
```diff
@@ -440,7 +444,7 @@ def _do_list(self, value):
     def _do_dict(self, value):
         idx = self._register(value)
         vals = [self._do(x) for x in value.values()]
-        data = dict(zip(value.keys(), vals))
+        data = dict(list(zip(list(value.keys()), vals)))
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/persistence/state_pickler.py]: neither of them are. corrected.

File: apptools/persistence/state_pickler.py [hunk 3]
```diff
@@ -270,7 +274,7 @@ def __init__(self):
                     }
         if PY_VER == 2:
             type_map[long] = self._do_basic_type
-            type_map[unicode] = self._do_basic_type
+            type_map[six.text_type] = self._do_basic_type
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/persistence/state_pickler.py]: yes but this should not have been touched as it is python 2-only code.

```
>>> import six
>>> six.text_type
<type 'unicode'>
```

File: apptools/appscripting/script_manager.py [hunk 4]
```diff
@@ -23,11 +24,12 @@
...
-from i_bind_event import IBindEvent
-from i_script_manager import IScriptManager
-from lazy_namespace import add_to_namespace, FactoryWrapper, LazyNamespace
-from scriptable_type import make_object_scriptable
+from .bind_event import BindEvent
+from .i_bind_event import IBindEvent
+from .i_script_manager import IScriptManager
+from .lazy_namespace import add_to_namespace, FactoryWrapper, LazyNamespace
+from .scriptable_type import make_object_scriptable
+import six
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply apptools/appscripting/script_manager.py]: Unfortunately modernize doesn't understand import order conventions :(

File: apptools/io/__init__.py [hunk 5]
```diff
@@ -14,4 +14,4 @@
 """ Provides an abstraction for files and folders in a file system.
     Part of the AppTools project of the Enthought Tool Suite.
 """
-from api import *
+from .api import *
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply apptools/io/__init__.py]: We could at least make it explicit imports.

File: apptools/lru_cache/tests/test_lru_cache.py [hunk 6]
```diff
@@ -141,7 +144,7 @@ def test_updated_event():
     c.on_trait_change(lambda x: events.append(x), 'updated')
 
     c[0] = 0
-    print(c.keys())
+    print(list(c.keys()))
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/lru_cache/tests/test_lru_cache.py]: removing. maybe it was a left over debugging statement.

File: apptools/undo/tests/test_command_stack.py [hunk 7]
```diff
@@ -12,14 +12,15 @@
...
 
-from __future__ import absolute_import
-from traits.testing.unittest_tools import unittest
 from contextlib import contextmanager
+
 from nose.tools import assert_equal
+from six.moves import range
 
+from traits.testing.unittest_tools import unittest
+from traits.testing.unittest_tools import unittest
```
Comment (jwiggins) [inline apptools/undo/tests/test_command_stack.py]: Duped.

File: apptools/io/h5/dict_node.py [hunk 8]
```diff
@@ -69,7 +69,7 @@ def __contains__(self, key):
         return key in self.data
 
     def keys(self):
-        return self.data.keys()
+        return list(self.data.keys())
```
Comment (rahulporuri) [inline apptools/io/h5/dict_node.py]: This change needs discussion.

If we don't make this change, we have different behavior on Python 2 and Python 3.
If wrap the call with `list`, the return value of `H5DictNode.keys` differs from how a normal dictionary behaves.

File: apptools/lru_cache/lru_cache.py [hunk 9]
```diff
@@ -98,15 +98,15 @@ def get(self, key, default=None):
 
     def items(self):
         with self._lock:
-            return self._cache.items()
+            return list(self._cache.items())
```
Comment (rahulporuri) [inline apptools/lru_cache/lru_cache.py]: Similar to an earlier question regd H5DictNode, should the items/keys/values methods return iterables on Python 3 or should they return lists?

File: apptools/io/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/io/__init__.py]: Made them full imports ("from apptools.io.api import *")

File: apptools/persistence/tests/test_state_pickler.py [hunk 10]
```diff
@@ -5,17 +5,24 @@
...
 import numpy
+import six
 
 from traits.api import Bool, Int, Long, Array, Float, Complex, Any, \
     Str, Unicode, Instance, Tuple, List, Dict, HasTraits
 
+if six.PY2:
+    long_trait_type = Long
+else:
+    long_trait_type = Int
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/persistence/tests/test_state_pickler.py]: Hmm. I checked and the Int and Long don't share a common baseclass on Python 3. But they do use the same/similar validator for traits on Python 3.

Either way, this code isn't necessary. It was added when I was debugging an issue that this PR added! Removed these changes.

File: apptools/io/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwiggins) [reply apptools/io/__init__.py]: I'm a little surprised that that doesn't cause a recursion error...

[the remark this answers is not in the document]
Comment (jwiggins) [reply apptools/io/__init__.py]: (also, I was more horrified by the `*` than the relative import)

File: apptools/io/h5/table_node.py [hunk 11]
```diff
@@ -1,6 +1,9 @@
+from __future__ import absolute_import
 import numpy as np
 from pandas import DataFrame
 from tables.table import Table as PyTablesTable
+import six
+from six.moves import zip
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/io/h5/table_node.py]: The builtin `zip` is adequate. I removed this.

File: apptools/io/h5/dict_node.py (hunk 8 above)
Comment (jwiggins) [reply apptools/io/h5/dict_node.py]: I think that because this method is part of the dictionary interface then it should act like a normal dictionary. Remove the `list` wrapper.

File: apptools/lru_cache/lru_cache.py (hunk 9 above)
Comment (jwiggins) [reply apptools/lru_cache/lru_cache.py]: I want to say they should return iterables, but the locking presents a thorny problem. Perhaps the safest solution would be to implement this as an iterator (explicitly `yield`-ing items) on Python 3, and using the old/existing implementation on Python 2.

Comment (rahulporuri) [reply apptools/lru_cache/lru_cache.py]: I'd prefer leaving this change as-is, mainly because we're interested in removing `LRUCache`. Is that okay?

**Comment by rahulporuri** [comment]
@jwiggins @corranwebster this PR is ready for review. again. for the last time, i thnk.

File: apptools/persistence/spickle.py [hunk 12]
```diff
@@ -15,6 +15,7 @@
 # Copyright (c) 2006-2015, Prabhu Ramachandran
 # License: BSD Style.
 
+from __future__ import absolute_import
 import sys
 if sys.version_info[0] > 2:
     raise ImportError("This module does not work with Python 3.x")
```
[the remark this answers is not in the document]
Comment (rahulporuri) [reply apptools/persistence/spickle.py]: I tried using the `exclude` kwarg when calling `find_packages` to exclude just this module from the Python 3 package but it looks like `exclude` will only help when entire packages/sub-packages need to be excluded and not when one specific module needs to be excluded.

File: apptools/io/h5/table_node.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 from pandas import DataFrame
 from tables.table import Table as PyTablesTable
+import six
```
Comment (jwiggins) [inline apptools/io/h5/table_node.py]: Move this up so that it's after the pandas import

File: apptools/io/h5/tests/test_file.py [hunk 14]
```diff
@@ -509,15 +509,15 @@ def test_attribute_iteration_methods():
...
         attrs['alpha'] = 0xff
 
-        items = attrs.items()
+        items = list(attrs.items())
 
         assert all(isinstance(x, tuple) for x in items)
 
         # unfold the pairs
-        keys, vals = map(list, zip(*items))
+        keys, vals = [list(item) for item in zip(*items)]
```
Comment (jwiggins) [inline apptools/io/h5/tests/test_file.py]: No idea why this was originally written in such a complicated way. Just write `keys, vals = zip(*items)`. It's a common Python idiom and everything.

File: apptools/lru_cache/tests/test_lru_cache.py [hunk 15]
```diff
@@ -73,7 +73,7 @@ def test_cache_resize():
 
 def test_cache_items():
     c = LRUCache(2)
-    assert_equal([], c.items())
```
Comment (jwiggins) [inline apptools/lru_cache/tests/test_lru_cache.py]: `LRUCache.items` returns a `list` already.

File: apptools/lru_cache/tests/test_lru_cache.py [hunk 16]
```diff
@@ -105,15 +105,15 @@ def test_cache_idempotency():
 
 def test_cache_keys_values():
     c = LRUCache(2)
-    assert_equal([], c.items())
+    assert_equal([], list(c.items()))
```
Comment (jwiggins) [inline apptools/lru_cache/tests/test_lru_cache.py]: `LRUCache.items` returns a `list` already.

File: apptools/naming/trait_defs/naming_traits.py [hunk 17]
```diff
@@ -27,6 +27,7 @@
 
 from apptools.naming.api \
     import Binding
+import six
```
Comment (jwiggins) [inline apptools/naming/trait_defs/naming_traits.py]: This should probably be moved to above the ETS imports

File: apptools/naming/ui/context_node_type.py [hunk 18]
```diff
@@ -15,11 +15,13 @@
 
 
 # Enthought library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (jwiggins) [reply apptools/naming/ui/context_node_type.py]: Also the `# Enthought library imports.` comment above this is a bit off

File: apptools/preferences/preferences.py [hunk 19]
```diff
@@ -1,8 +1,8 @@
 """ The default implementation of a node in a preferences hierarchy. """
 
-from __future__ import print_function
-
 # Standard library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (jwiggins) [reply apptools/preferences/preferences.py]: Useless diff now

File: apptools/preferences/tests/preferences_test_case.py [hunk 20]
```diff
@@ -164,14 +164,14 @@ def test_keys(self):
...
+        self.assertEqual([], list(p.keys()))
 
         # Set some preferences in the node.
         p.set('a', '1')
         p.set('b', '2')
         p.set('c', '3')
 
-        keys = p.keys()
+        keys = list(p.keys())
         keys.sort()
```
Comment (jwiggins) [inline apptools/preferences/tests/preferences_test_case.py]: One liner: `keys = sorted(p.keys())`

File: apptools/preferences/ui/preferences_node.py [hunk 21]
```diff
@@ -15,11 +15,13 @@
 
 
 # Enthought library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (jwiggins) [reply apptools/preferences/ui/preferences_node.py]: Not an Enthought library import

File: apptools/scripting/recorder.py [hunk 22]
```diff
@@ -221,9 +221,9 @@ def register(self, object, parent=None, trait_name_on_parent='',
             # Always ignore these.
             ignore.extend(['trait_added', 'trait_modified'])
 
-            sub_recordables = object.traits(record=True).keys()
+            sub_recordables = list(object.traits(record=True).keys())
             # Find all the trait names we must ignore.
-            ignore.extend(object.traits(record=False).keys())
+            ignore.extend(list(object.traits(record=False).keys()))
```
Comment (jwiggins) [inline apptools/scripting/recorder.py]: You can pass an iterable to `list.extend`

File: apptools/sweet_pickle/tests/updater_test_case.py [hunk 23]
```diff
@@ -238,8 +238,8 @@ def fn2():
         self.assertEqual(2, len(self.registry.state_functions))
         self.assertEqual(2, len(self.registry.version_attribute_map))
         self.assertEqual(1, len(self.registry._state_function_classes))
-        self._validate_class_map_contents(updater.class_map.items() + \
-            updater2.class_map.items())
+        self._validate_class_map_contents(list(updater.class_map.items()) + \
```
Comment (jwiggins) [inline apptools/sweet_pickle/tests/updater_test_case.py]: The backtick is unnecessary

File: apptools/io/h5/tests/test_file.py (hunk 14 above)
Comment (mdickinson) [reply apptools/io/h5/tests/test_file.py]: FWIW, I'm really not a fan of the `zip(*stuff)` idiom: it's got a nasty not-immediately-obvious corner case where `stuff` is empty. Fine in this case, but then it gets re-used in places where it's not fine and people don't notice the corner case until it's too late.

```
>>> d = {'a': 1, 'b': 2}
>>> keys, values = zip(*d.items())  # fine
>>> d = {}
>>> keys, values = zip(*d.items())  # not fine
```

Comment (mdickinson) [reply apptools/io/h5/tests/test_file.py]: This doesn't detract from @jwiggins' comment, of course.

One possible reason for the `list` is that otherwise `keys` and `values` are tuples. I haven't looked at the context, so I don't know if that matters in this particular case, but that's the usual reason for the extra `map(list, ...)` construct.

Comment (rahulporuri) [reply apptools/io/h5/tests/test_file.py]: `keys`, `vals` end up as tuples when we just use `= zip(*items)`, which is not expected as we then try comparing to lists. I'm keeping the `[list(item) ... ]` code instead of updating other code to test against tuples or wrap `keys`/`vals` with `list` later in the code.

Comment (jwiggins) [reply apptools/io/h5/tests/test_file.py]: Now that you've reminded me that `zip` returns tuples, that explains why it was initially written as `map(list, zip(*items))`.

**Review by jwiggins — APPROVED** [review]
